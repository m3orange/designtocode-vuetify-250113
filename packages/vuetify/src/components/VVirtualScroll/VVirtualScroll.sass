.v-virtual-scroll
  display: block
  height: 100%
  overflow: auto
  position: relative

  &__container
    display: block

  &__item
    position: absolute
    left: 0
    right: 0
