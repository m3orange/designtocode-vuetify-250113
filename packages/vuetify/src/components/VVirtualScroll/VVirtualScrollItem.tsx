// Composables
import { makeComponentProps } from '@/composables/component'
import { useResizeObserver } from '@/composables/resizeObserver'
import { useToggleScope } from '@/composables/toggleScope'

// Utilities
import { onUpdated, watch } from 'vue'
import { genericComponent, propsFactory, useRender } from '@/util'
import { useVirtualItem } from '@/composables/virtual'

export const makeVVirtualScrollItemProps = propsFactory({
  virtualKey: {
    type: [String, Number],
    required: true,
  },
  dynamicHeight: Boolean,
  renderless: Boolean,

  ...makeComponentProps(),
}, 'VVirtualScrollItem')

export const VVirtualScrollItem = genericComponent<{ default: { props: Record<string, any> } | void }>()({
  name: 'VVirtualScrollItem',

  props: makeVVirtualScrollItemProps(),

  setup (props, { emit, slots }) {
    const { handleItemResize } = useVirtualItem()
    const { resizeRef, contentRect } = useResizeObserver()

    useToggleScope(() => props.dynamicHeight, () => {
      watch(() => contentRect.value?.height, height => {
        if (height != null) handleItemResize(props.virtualKey, height)
      })
    })

    function updateHeight () {
      if (props.dynamicHeight && contentRect.value) {
        handleItemResize(props.virtualKey, contentRect.value.height)
      }
    }

    onUpdated(updateHeight)

    useRender(() => props.renderless ? (
      <>
        { slots.default?.({ props: { ref: props.dynamicHeight ? resizeRef : undefined, key: props.virtualKey } }) }
      </>
    ) : (
      <div
        ref={ props.dynamicHeight ? resizeRef : undefined }
        class={[
          'v-virtual-scroll__item',
          props.class,
        ]}
        style={ props.style }
      >
        { slots.default?.() }
      </div>
    ))
  },
})
