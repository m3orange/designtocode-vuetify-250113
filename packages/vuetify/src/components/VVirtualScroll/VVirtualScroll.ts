import './VVirtualScroll.sass'

// Types
import Vue, { VNode } from 'vue'

export default Vue.extend({
  name: 'v-virtual-scroll',

  props: {
    height: {
      type: String,
      default: '100%',
    },
    itemHeight: {
      type: Number,
      required: true,
    },
    items: {
      type: Array,
      required: true,
    },
  },

  data: () => ({
    bench: 5,
    first: 0,
    last: 0,
    scrollTop: 0,
  }),

  computed: {
    firstToRender (): number {
      return Math.max(0, this.first - this.bench)
    },
    lastToRender (): number {
      return Math.min(this.items.length, this.last + this.bench)
    },
  },

  mounted (): void {
    this.last = this.getLast(0)
    this.$el.addEventListener('scroll', this.onScroll, false)
  },

  beforeDestroy (): void {
    this.$el.removeEventListener('scroll', this.onScroll, false)
  },

  methods: {
    getChildrenVNodes (h: Function): VNode[] {
      const vNodes: any[] = this.items.slice(this.firstToRender, this.lastToRender)
      return vNodes.map((item: any, i: number): VNode => h('div', {
        staticClass: 'v-virtual-scroll__item',
        style: {
          top: (this.firstToRender + i) * this.itemHeight + 'px',
        },
      }, this.$scopedSlots.default && this.$scopedSlots.default(item)))
    },
    getFirst (): number {
      return Math.floor(this.scrollTop / this.itemHeight)
    },
    getLast (first: number): number {
      return first + Math.ceil(this.$el.clientHeight / this.itemHeight)
    },
    onScroll (e: Event): void {
      const target = e.currentTarget as HTMLElement
      this.scrollTop = target.scrollTop
      this.first = this.getFirst()
      this.last = this.getLast(this.first)
    },
  },

  render (h: Function): VNode {
    const childrenVNodes: VNode[] = this.getChildrenVNodes(h)
    const parentVNode: VNode = h('div', {
      staticClass: 'v-virtual-scroll__container',
      style: {
        height: (this.items.length * this.itemHeight) + 'px',
      },
    }, childrenVNodes)
    const scrollVNode: VNode = h('div', {
      staticClass: 'v-virtual-scroll',
      style: {
        height: this.height,
      },
    }, [parentVNode])
    return scrollVNode
  },
})
