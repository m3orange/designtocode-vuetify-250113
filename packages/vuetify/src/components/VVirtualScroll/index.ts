// Styles
import './VVirtualScroll.sass'

import VVirtualScroll from './VVirtualScroll'

export { VVirtualScroll }
export default VVirtualScroll
