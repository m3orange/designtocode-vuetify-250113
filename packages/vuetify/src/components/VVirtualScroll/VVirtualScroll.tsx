// Styles
import './VVirtualScroll.sass'

// Components
import { VVirtualScrollItem } from './VVirtualScrollItem'

// Composables
import { makeComponentProps } from '@/composables/component'
import { makeDimensionProps, useDimension } from '@/composables/dimensions'
import { makeVirtualProps, useVirtual } from '@/composables/virtual'

// Utilities
import { toRef } from 'vue'
import {
  convertToUnit,
  genericComponent,
  propsFactory,
  useRender,
} from '@/util'

// Types
import type { PropType } from 'vue'
import type { GenericProps } from '@/util'

export interface VVirtualScrollSlot<T> {
  item: T
  index: number
}

export const makeVVirtualScrollProps = propsFactory({
  items: {
    type: Array as PropType<readonly unknown[]>,
    default: () => ([]),
  },

  ...makeVirtualProps(),
  ...makeComponentProps(),
  ...makeDimensionProps(),
}, 'VVirtualScroll')

export const VVirtualScroll = genericComponent<new <T>(
  props: {
    items?: readonly T[]
  },
  slots: {
    default: VVirtualScrollSlot<T>
  }
) => GenericProps<typeof props, typeof slots>>()({
  name: 'VVirtualScroll',

  props: makeVVirtualScrollProps(),

  setup (props, { slots }) {
    const { dimensionStyles } = useDimension(props)
    const {
      containerRef,
      handleScroll,
      scrollToIndex,
      paddingTop,
      paddingBottom,
      virtualItems,
    } = useVirtual(props, toRef(props, 'items'))

    useRender(() => (
      <div
        ref={ containerRef }
        class={[
          'v-virtual-scroll',
          props.class,
        ]}
        onScroll={ handleScroll }
        style={[
          dimensionStyles.value,
          props.style,
        ]}
      >
        <div
          class="v-virtual-scroll__container"
          style={{
            paddingTop: convertToUnit(paddingTop.value),
            paddingBottom: convertToUnit(paddingBottom.value),
          }}
        >
          { virtualItems.value.map((item, index) => (
            <VVirtualScrollItem
              key={ item.key }
              virtualKey={ item.key }
              dynamicHeight={ !props.itemHeight }
            >
              { slots.default?.({ item, index }) }
            </VVirtualScrollItem>
          ))}
        </div>
      </div>
    ))

    return {
      scrollToIndex,
    }
  },
})

export type VVirtualScroll = InstanceType<typeof VVirtualScroll>
