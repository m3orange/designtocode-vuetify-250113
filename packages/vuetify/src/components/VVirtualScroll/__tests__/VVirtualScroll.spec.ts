// Component
import VVirtualScroll from '../VVirtualScroll'

// Utilities
import {
  shallowMount,
  Wrapper,
} from '@vue/test-utils'

describe('VVirtualScroll.ts', () => {
  type Instance = InstanceType<typeof VVirtualScroll>
  let mountFunction: (options?: object) => Wrapper<Instance>
  let propsData: Object
  let mock: jest.SpyInstance
  const elementHeight: number = 100

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return shallowMount(VVirtualScroll, {
        ...options,
        scopedSlots: {
          default (item) {
            return this.$createElement('div', { class: 'item' }, item)
          },
        },
      })
    }
    propsData = {
      height: `${elementHeight}px`,
      items: [1, 2, 3],
      itemHeight: 50,
    }

    // mock clientHeight
    mock = jest.spyOn(window.HTMLElement.prototype, 'clientHeight', 'get').mockReturnValue(elementHeight)
  })

  afterEach(() => {
    mock.mockRestore()
  })

  it('should render component with scopedSlot and match snapshot', () => {
    const wrapper = mountFunction({
      propsData,
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should set height of scrollable element', () => {
    const wrapper = mountFunction({
      propsData,
    })

    const scrollable = wrapper.find('.v-virtual-scroll__container')
    expect(scrollable.element.style.height).toEqual('150px')
  })

  it('should render not more than 5 hidden items and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        height: `${elementHeight}`,
        items: [1, 2, 3, 4, 5, 6, 7, 8, 9, 0],
        itemHeight: 50,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render right items on scroll and match snapshot', async () => {
    const wrapper = mountFunction({
      propsData: {
        height: `${elementHeight}px`,
        items: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
        itemHeight: 50,
      },
    })

    const evtPayload = new CustomEvent('scroll', { detail: { currentTarget: { scrollTop: 500 } } })
    wrapper.vm.onScroll(evtPayload.detail as unknown as Event) // can't make ts work here

    expect(wrapper.html()).toMatchSnapshot()
  })
})
