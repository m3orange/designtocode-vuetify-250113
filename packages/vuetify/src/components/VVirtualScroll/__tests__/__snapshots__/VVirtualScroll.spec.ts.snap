// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VVirtualScroll.ts should render component with scopedSlot and match snapshot 1`] = `
<div class="v-virtual-scroll"
     style="height: 100px;"
>
  <div class="v-virtual-scroll__container"
       style="height: 150px;"
  >
    <div class="v-virtual-scroll__item"
         style="top: 0px;"
    >
      <div class="item">
        1
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 50px;"
    >
      <div class="item">
        2
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 100px;"
    >
      <div class="item">
        3
      </div>
    </div>
  </div>
</div>
`;

exports[`VVirtualScroll.ts should render not more than 5 hidden items and match snapshot 1`] = `
<div class="v-virtual-scroll">
  <div class="v-virtual-scroll__container"
       style="height: 500px;"
  >
    <div class="v-virtual-scroll__item"
         style="top: 0px;"
    >
      <div class="item">
        1
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 50px;"
    >
      <div class="item">
        2
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 100px;"
    >
      <div class="item">
        3
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 150px;"
    >
      <div class="item">
        4
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 200px;"
    >
      <div class="item">
        5
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 250px;"
    >
      <div class="item">
        6
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 300px;"
    >
      <div class="item">
        7
      </div>
    </div>
  </div>
</div>
`;

exports[`VVirtualScroll.ts should render right items on scroll and match snapshot 1`] = `
<div class="v-virtual-scroll"
     style="height: 100px;"
>
  <div class="v-virtual-scroll__container"
       style="height: 1000px;"
  >
    <div class="v-virtual-scroll__item"
         style="top: 250px;"
    >
      <div class="item">
        6
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 300px;"
    >
      <div class="item">
        7
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 350px;"
    >
      <div class="item">
        8
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 400px;"
    >
      <div class="item">
        9
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 450px;"
    >
      <div class="item">
        10
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 500px;"
    >
      <div class="item">
        11
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 550px;"
    >
      <div class="item">
        12
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 600px;"
    >
      <div class="item">
        13
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 650px;"
    >
      <div class="item">
        14
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 700px;"
    >
      <div class="item">
        15
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 750px;"
    >
      <div class="item">
        16
      </div>
    </div>
    <div class="v-virtual-scroll__item"
         style="top: 800px;"
    >
      <div class="item">
        17
      </div>
    </div>
  </div>
</div>
`;
